src/compressor.rs

```diff
@@ -332,50 +332,440 @@ mod level_tests {
+        let mut initial: BTreeMap<i64, StateGroupEntry> = BTreeMap::new();
+        let mut prev = None;
+
+        // This starts with the following structure
+        //
+        // (note missing 3-4 link)
+        // 0-1-2-3
+        // 4-5-6-7-8-9-10-11-12-13
+        //
+        // Each group i has state:
+        //     ('node','is',      i)
+        //     ('group',  j, 'seen') where j is ancestor of i
+        for i in 0i64..=13i64 {
+            if i == 4 {
+                prev = None
+            }
+            let mut entry = StateGroupEntry {
                 in_range: true,
                 prev_state_group: prev,
                 state_map: StateMap::new(),
-            },
-        );
+            };
+            entry
+                .state_map
+                .insert("group", &i.to_string(), "seen".into());
+            entry.state_map.insert("node", "is", i.to_string().into());
 
-        prev = Some(i)
-    }
+            initial.insert(i, entry);
 
-    let compressor = Compressor::compress(&initial, &[3, 3]);
-
-    let new_state = compressor.new_state_group_map;
-
-    let expected_edges: BTreeMap<i64, i64> = vec![
-        (1, 0),
-        (2, 1),
-        (4, 3),
-        (5, 4),
-        (6, 3),
-        (7, 6),
-        (8, 7),
-        (9, 6),
-        (10, 9),
-        (11, 10),
-        (13, 12),
-    ]
-    .into_iter()
-    .collect();
-
-    for sg in 0i64..=13i64 {
-        assert_eq!(
-            expected_edges.get(&sg).cloned(),
-            new_state[&sg].prev_state_group,
-            "state group {} did not match expected",
-            sg,
-        );
+            prev = Some(i)
+        }
+
+        let mut compressor = Compressor {
+            original_state_map: &initial,
+            new_state_group_map: BTreeMap::new(),
+            levels: vec![Level::new(3), Level::new(3)],
+            stats: Stats::default(),
+        };
+        compressor.create_new_tree();
+
+        let new_state = &compressor.new_state_group_map;
+
+        for n in new_state {
+            println!("{:?}", n);
+        }
+
+        // This should create the following structure
+        //
+        // Brackets mean that has NO predecessor but is in that position in the
+        // levels tree
+        //
+        // 0  3\        12
+        // 1 (4)(6)\    13
+        // 2  5  7  9
+        //       8  10
+        //          11
+        let expected_edges: BTreeMap<i64, i64> = vec![
+            (1, 0),
+            (2, 1),
+            (5, 4),
+            (7, 6),
+            (8, 7),
+            (9, 6),
+            (10, 9),
+            (11, 10),
+            (13, 12),
+        ]
+        .into_iter()
+        .collect();
+
+        for sg in 0i64..=13i64 {
+            assert_eq!(
+                expected_edges.get(&sg).cloned(),
+                new_state[&sg].prev_state_group,
+                "state group {} did not match expected",
+                sg,
+            );
+        }
```

Comment:
This file is now getting quite long, I think maybe we should move such tests out into the `tests/` folder?

Reply:
The trouble with moving them into `tests/` is that I then can only use the things that are public - it is getting a bit long though!

Reply:
I could put them in seperate files by moving them into a `compressor/` folder:

So `src/compressor.rs` would have:
```rust
#[cfg(test)]
mod compressor_tests;
```

And then there would be a file `src/compressor/compressor_tests` that had
```rust
use xyz;

#[test]
fn test_this_does_that() {
    ...
}
````

would you be happy with that?

Reply:
wfm!